src/environmentd/src/http/webhook.rs

```diff
@@ -0,0 +1,306 @@
+        .append_webhook(database.clone(), schema.clone(), name.clone())
+        .await?;
+    let Some((appender, body_ty, header_ty)) = maybe_appender else {
+        let desc = format!("{database}.{schema}.{name}");
+        return Err(WebhookError::NotFound(desc.quoted().to_string()));
+    };
+
+    // Pack our body and headers into a Row.
+    let row = pack_row(body, headers, body_ty, header_ty)?;
+
+    // Send the row to get appended.
+    appender.append(vec![(row, 1)]).await?;
+
+    Ok(())
+}
+
+/// Given the body and headers of a request, pack them into a [`Row`].
+fn pack_row(
+    body: Bytes,
+    headers: http::HeaderMap,
+    body_ty: ColumnType,
+    header_ty: Option<ColumnType>,
+) -> Result<Row, WebhookError> {
+    // If we're including headers then we have two columns.
+    let num_cols = header_ty.as_ref().map(|_| 2).unwrap_or(1);
+
+    // Pack our row.
+    let mut row = Row::with_capacity(num_cols);
+    let mut packer = row.packer();
+
+    // Pack our body into a row.
+    match body_ty.scalar_type {
+        ScalarType::Bytes => packer.push(Datum::Bytes(&body[..])),
+        ty @ ScalarType::String => {
+            let s = std::str::from_utf8(&body).map_err(|m| WebhookError::InvalidBody {
+                ty,
+                msg: m.to_string(),
+            })?;
+            packer.push(Datum::String(s));
+        }
+        ty @ ScalarType::Jsonb => {
+            let jsonb_packer = JsonbPacker::new(&mut packer);
+            jsonb_packer
+                .pack_slice(&body[..])
+                .map_err(|m| WebhookError::InvalidBody {
+                    ty,
+                    msg: m.to_string(),
+                })?;
+        }
+        ty => {
+            Err(anyhow::anyhow!(
+                "Invalid body type for Webhook source: {ty:?}"
+            ))?;
+        }
+    }
+
+    // Pack our headers into a row, if we're supposed to.
+    let mut headers_s = BTreeMap::new();
+    for (name, val) in headers.iter() {
+        if let Ok(val_s) = val.to_str() {
+            headers_s.insert(name.as_str(), val_s);
+        }
+    }
+    if header_ty.is_some() {
+        packer.push_dict(
+            headers_s
+                .iter()
+                .map(|(name, val)| (*name, Datum::String(*val))),
+        );
+    }
+
+    Ok(row)
+}
+
+/// Errors we can encounter when appending data to a Webhook Source.
+///
+/// Webhook sources are a bit special since they are handled by `environmentd` (all other sources
+/// are handled by `clusterd`) and data is "pushed" to them (all other source pull data). The
+/// errors also generally need to map to HTTP status codes that we can use to respond to a webhook
+/// request. As such, webhook errors don't cleanly map to any existing error type, hence the
+/// existence of this error type.
+#[derive(Error, Debug)]
+pub enum WebhookError {
+    #[error("no object was found at the path {}", .0.quoted())]
+    NotFound(String),
+    #[error("a resource was exhausted: {0}")]
```

Comment:
This isn't a useful string for a user. It should change or move to `InternalStorageError`.

Reply:
Sounds good, removed!